src/Component.php

```diff
@@ -44,32 +35,37 @@ private function processDir(Finder $finder, string $dir, string $filter) : void
                 // copy manifest without modification
                 $fs->copy($inFile->getPathname(), $this->getDataDir() . "/out/$dir/" . $inFile->getFilename());
             } else {
-                try {
-                    $source = fopen($inFile->getPathname(), 'r');
-                    $destinationDir = $this->getDataDir() . "/out/$dir/" . $inFile->getRelativePath();
-                    $fs->mkdir($destinationDir);
-                    $destination = fopen(
-                        $destinationDir . '/' . $inFile->getFilename(),
-                        'w'
-                    );
-                } /** @noinspection PhpRedundantCatchClauseInspection */ catch (\ErrorException $e) {
-                    throw new \Exception('Failed opening in/out ' . $dir . ' ' . $e->getMessage());
-                }
+                $destinationDir = $this->getDataDir() . "/out/$dir/" . $inFile->getRelativePath();
+                $fs->mkdir($destinationDir);
+                $destinationFile = $destinationDir . '/' . $inFile->getFilename();
+                $this->processFile($inFile, $destinationFile, $filter);
+            }
+        }
+    }
 
-                stream_filter_append($destination, $filter, STREAM_FILTER_WRITE);
-                $converted = '';
-                try {
-                    while (!feof($source)) {
-                        $converted = fread($source, 1000);
-                        fwrite($destination, $converted);
-                    }
-                } /** @noinspection PhpRedundantCatchClauseInspection */ catch (\ErrorException $e) {
-                    throw new UserException('Conversion failed in ' . $inFile->getFilename() . '. ' .
-                        $e->getMessage() . '. Failing text: ' . $converted);
+    private function processFile(SplFileInfo $inFile, string $destinationFileName, string $filter) : void
+    {
+        try {
+            $source = fopen($inFile->getPathname(), 'r');
+            $destination = fopen($destinationFileName, 'w');
+        } /** @noinspection PhpRedundantCatchClauseInspection */ catch (\ErrorException $e) {
+            throw new \Exception('Failed opening in/out ' . $inFile->getPathname() . ' ' . $e->getMessage());
+        }
+        try {
+            stream_filter_append($destination, $filter, STREAM_FILTER_WRITE);
+            $converted = '';
+            try {
+                while (!feof($source)) {
+                    $converted = fread($source, 1000);
```

Comment:
ztratil se můj komentář o stream copy to stream. Osobně totiž nevím, jak se chová ten byte limit u fread chová s multibyte kódováním a zavání to nějakou těžko odhalitelnou chybou, která se objeví jen pokud na 100tém byte je multibyte a přesně to vyjde s předchozím textem. Co když to vyjde tak, že ho to zařízne v půlce multi-byte písmena nebo tak něco ([protože utf může mít 1-6B](https://stackoverflow.com/a/33349765/112000))?

Reply:
no na ten jsem odpovidal timhle http://php.net/manual/en/function.stream-copy-to-stream.php#98119
to s tim 100b me samozrejme napadalo, a zkousel jsem to - kupodivu to funguje i s fread($source, 1); a to dokonce tak, ze v $converted je pulka znaku, ale ten streamfilter ji nezapise dokud neprijde druha, vypada to, ze to neni uplne blby

Reply:
7 years ago, PHP 5.1, 0 upvotes, bral bych to s rezervou. Ale cool, že jsi to otestoval, tím pádem OK

Reply:
Pro zajímavost:
Implementace:
https://github.com/php/php-src/blob/a7fe2570d3ce6915d4ea85c62c0f880ddc225ba7/main/streams/streams.c#L1462
Buďto to dělá přes mmap pokud to jde nebo to tahá po CHUNK_SIZE a ten je nejspíš definovaný 8192. 
https://github.com/php/php-src/blob/a7fe2570d3ce6915d4ea85c62c0f880ddc225ba7/main/streams/php_streams_int.h#L49

Nicméně protože to máme filtrované a ten check na mmap je implementovaný takhle: https://github.com/php/php-src/blob/a7fe2570d3ce6915d4ea85c62c0f880ddc225ba7/main/streams/php_stream_mmap.h#L67

tak to pojede vždycky tím 8192 bufferem v tomhle případě.

Reply:
no nedalo me to, tak jsem se v tom jeste povrtal, takze
`stream_copy_to_stream($source, $destination)` skutecne da `PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 660698201 bytes)`
alternativa je pouzit
```
                $offset = 0;
                $bytes = true;
                while ($bytes) {
                    $bytes = stream_copy_to_stream($source, $destination, 1024, $offset);
                    $offset = $offset + 1024;
                }
```
akoratze teda prijdu o vyhodu zobrazit tu cast kde to failnulo a navic je to pomalejsi ~46s vs ~36s (fread/fwrite) coz me samotnyho prekvapilo (nicmene vypada to, ze je to 10s vzdycky a nezavisi to a mnostvi dat). Krome toho jsem zkousel i 8192 buffer - i tam se to konvertuje spravne i kdyz vyjde pulka znaku zrovna do konce bufferu. Takze si to apruvuju taky, protoze jsme s temahle peti radkma stravili silene energie :) Ale aspon jsme o neco chytejsi.

Reply:
Data driven. Hezky. Díky.